AF-2811: Required Process Variable Tags shouldn't be accepted in forms when empty

akumar074 opened the pull request:
* Added "required" and "readonly" property to FormField by tags
 * Fixed validation for input box in ProcessInstance form

**JIRA**: 
[AF-2811](https://issues.redhat.com/browse/AF-2811)
[RHPAM-3047](https://issues.redhat.com/browse/RHPAM-3047)

**Referenced Pull Requests**:
* https://github.com/kiegroup/kie-wb-common/pull/3592
* https://github.com/kiegroup/jbpm-wb/pull/1484

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

akumar074 commented:
Jenkins run fdb

akumar074 commented:
Jenkins retest this

akumar074 commented:
Jenkins retest this

tomasdavidorg commented:
Hello @akumar074 ,
I tried the fix but I found an issue:
1. Submit empty required input  -> Error dialog.
2. Close Error dialog, add value to the input, submit -> The same error dialog.

The error message will be unclear to users. Would it be possible to improve the meaning of the message?
The input does not have the * (Asterisk). Shouldn't it be there? 
Would it be also possible to add a test to this PR?

elguardian commented:
Hi @akumar074 
we are losing information about the tags here. Why this line is removed ?
https://github.com/kiegroup/droolsjbpm-integration/pull/2398/files#diff-87c87b4946c84e3b136dba324361a55b86d96a3e2b3a9b5b09f528980bbb19a3R262-R263

You are restricting the use case of this. You might not be interested but the idea is to make available the information to anyone. 
You can compute that check in the UI (Read only, required)

akumar074 commented:
Jenkins run fdb

akumar074 commented:
@tomasdavidorg I fixed the issue and updated the PR.
@elguardian Thanks for the heads up. I changed in client and fixed the validation problem in the input box.

tomasdavidorg commented:
Hi @akumar074 , I still can see the issue.
1. I created a simple process with a single required process variable.
2. I started the process (without value) and I got the error.
3. After that I closed the error popup and I added some value and the error is still there.
![image](https://user-images.githubusercontent.com/4498639/109958378-73ad6d80-7ce6-11eb-8f89-3ce367d823fb.png)

akumar074 commented:
Jenkins run fdb

akumar074 commented:
Jenkins run fdb

yesamer commented:
jenkins retest this please

akumar074 commented:
Jenkins run fdb

tomasdavidorg reviewed (approved):
Looks good to me. Just the two comments about formatting. 🙂

ederign commented:
@pefernan could you please take a look at this?

pefernan commented on kie-server-parent/kie-server-services/kie-server-services-jbpm-ui/src/main/java/org/kie/server/services/jbpm/ui/form/render/AbstractFormRenderer.java:
```diff
@@ -449,9 +440,14 @@ protected void processFormLayout(FormInstance topLevelForm,
-                            item.setReadOnly(field.isReadOnly());
-                            item.setRequired(field.isRequired());
+                            Set<String> tags = field.getTags();
+                            if(tags != null) {
```
does this mean that tags override any config on the form definition??

akumar074 replied:
Hi @pefernan , yes, the tags may have required and read-only property from process data for a field.

pefernan replied:
My only concern here would be that if a field is marked as required or readonly on the form and the field has an empty set of tags, it wil always be set to non-required/readonly. I'm not saying that's wrong, just pointing a possible issue.

Are the tags also taken into account on during the form generation process? Because that would be another interesting point.

pefernan reviewed (changes requested):
After testing it, looks like this change only affects to HTML forms rendered by kie-server and the auto generated forms (if no form is present). I'd add some testing but a part from that I think that's ok.

@akumar074 thanks!

akumar074 commented:
> After testing it, looks like this change only affects to HTML forms rendered by kie-server and the auto generated forms (if no form is present). I'd add some testing but a part from that I think that's ok.
> 
> @akumar074 thanks!

@pefernan Thanks for the review, I have updated the tests and will trigger the build.

akumar074 commented:
Jenkins retest this

akumar074 commented:
Jenkins run fdb

akumar074 commented:
Jenkins retest this

pefernan reviewed (approved):
+1, thank you @akumar074 when green.

akumar074 commented:
Jenkins run fdb

akumar074 commented:
Jenkins run fdb

akumar074 commented:
Jenkins retest this

akumar074 commented:
Jenkins run fdb

tomasdavidorg commented:
jenkins do fdb

tomasdavidorg commented:
@akumar074 I can see same issue as on 7.52.x branch https://github.com/kiegroup/droolsjbpm-integration/pull/2450#issuecomment-817961426

akumar074 commented:
Jenkins run fdb

tomasdavidorg commented:
> @akumar074 I can see same issue as on 7.52.x branch [#2450 (comment)](https://github.com/kiegroup/droolsjbpm-integration/pull/2450#issuecomment-817961426)

Issue was reported https://issues.redhat.com/browse/RHPAM-3595

akumar074 commented:
Jenkins retest this

akumar074 commented:
Jenkins run fdb

akumar074 commented:
Jenkins run fdb
Jenkins retest this

akumar074 commented:
Jenkins run fdb

akumar074 commented:
Jenkins run fdb

akumar074 commented:
jenkins run fdb

akumar074 commented:
jenkins retest this

paulovmr commented:
jenkins run fdb
